src/prog_models/utils/traj_gen/trajectory.py

```diff
@@ -0,0 +1,624 @@
+
+    :param phi:         double, n x 1, first Euler's angle as a function of time.
+    :param theta:       double, n x 1, second Euler's angle as a function of time.
+    :param psi:         double, n x 1, third Euler's angle as a function of time.
+    :param delta_t:     double, scalar, time step size, (default=1).
+    :return p:          double, n x 1, body roll rate as a function of time
+    :return q:          double, n x 1, body pitch rate as a function of time
+    :return r:          double, n x 1, body yaw rate as a function of time
+    """
+    phidot   = np.insert(np.diff(phi) / delta_t, 0, 0.0)
+    thetadot = np.insert(np.diff(theta) / delta_t, 0, 0.0)
+    psidot   = np.insert(np.diff(psi) / delta_t, 0, 0.0)
+    p        = np.zeros_like(phidot)
+    q        = np.zeros_like(phidot)
+    r        = np.zeros_like(phidot)
+    for ii in range(len(phi)):
+        des_angular_vel = geom.body_ang_vel_from_eulers(phi[ii], theta[ii], psi[ii], phidot[ii], thetadot[ii], psidot[ii])
+        p[ii], q[ii], r[ii] = des_angular_vel[0], des_angular_vel[1], des_angular_vel[2]
+    return p, q, r
+
+
+def derivate_position(p, dt):
+    """
+    Given a position profile as a function of time, p = p(t), derivate it up to 3 times to obtain velocity, acceleration, and jerk
+
+    For simplicity of computation as part of a trajectory profile, the derivatives are force to start at 0, e.g.,  velocity[0] = 0.0
+
+    :param p:           n x 1 array, doubles, position profile as a function of time
+    :param dt:          scalar, double, time step size used to generate the position profile, which also serves as derivative increment
+    :return:            three n x 1 arrays of doubles in a list, corresponding to velocity, acceleration and jerk, respectively.
+    """
+    v = np.zeros_like(p)
+    a = np.zeros_like(p)
+    j = np.zeros_like(p)
+
+    v = np.gradient(p, dt)
+    v[0] = 0.
+    a = np.gradient(v, dt)
+    a[0] = 0.
+    j = np.gradient(j, dt)
+    j[0] = 0.
+    return v, a, j
+
+
+def gen_attitude(psi, ax, ay, az, max_phi, max_theta, gravity):
+    # --------- Calculate angular kinematics based on acceleration and yaw ---------- #
+    # linearized angular kinematics
+    phi   = 1.0 / (gravity + az) * (ax * np.sin(psi) - ay * np.cos(psi))    # 
+    theta = 1.0 / (gravity + az) * (ax * np.cos(psi) + ay * np.sin(psi))    # 
+    
+    # Introduce limits on attitude angles
+    phi   = np.fmax(np.fmin(phi, max_phi), -max_phi)
+    theta = np.fmax(np.fmin(theta, max_theta), -max_theta)
+    return phi, theta, psi
+   
+class Trajectory():
+    """
+    Trajectory class.
+
+    This class generates a trajectory object that is used to create a smooth profile including
+    position, velocity, acceleration, Euler's angles, and attitude rates.
+    The profiles are used to generate the desired state of a vehicle that needs to be followed to complete the trajectory.
+    
+    Input variables:
```

Comment:
This needs to be updated to conform with numpy or google docstring format to work with sphinx